### PR TITLE
#29532: Update git pre-push hook so that only upstream branches can get pushed to origin

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/29532

### CODE REVIEW

File: scripts/maint/pre-push.git-hook [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
 
 z40=0000000000000000000000000000000000000000
 
+remote="$1"
 CUR_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+# Only allow pushing master, release-* and maint-* branches to origin.
+if [ "$remote" == "origin" ]
+then
+        if  [ "$CUR_BRANCH" != "master" ] || [[ $CUR_BRANCH != release-* ]] ||
```
Comment (nmathewson) [inline scripts/maint/pre-push.git-hook]: Should these `||` operators be `&&` ?

File: scripts/maint/pre-push.git-hook [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
-# This is git pre-push hook script to prevent "fixup!" and "squash!" commits
-# from ending up in upstream branches (master, release-* or maint-*).
-#
 # The following sample script was used as starting point:
 # https://github.com/git/git/blob/master/templates/hooks--pre-push.sample
 
 z40=0000000000000000000000000000000000000000
 
+remote="$1"
 CUR_BRANCH=$(git rev-parse --abbrev-ref HEAD)
```
Comment (nmathewson) [inline scripts/maint/pre-push.git-hook]: I know this line isn't new, but does it still work if you're pushing something other than HEAD, as in "git push branchname reponame" ?

File: scripts/maint/pre-push.git-hook (hunk 1 above)
Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: Yes, fixed in fb6e9ba26dc411ac79d0b6d3dabf71868ba83de5.

File: scripts/maint/pre-push.git-hook (hunk 2 above)
Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: Good point. I tried to make it more general in 60c8af367799e2cb8f59a236e97fdb0ad04bc350.

File: scripts/maint/pre-push.git-hook [hunk 3]
```diff
@@ -52,6 +43,19 @@ do
 			range="$remote_sha..$local_sha"
 		fi
 
+                if ref_is_upstream_branch "$local_ref" == 0 ||
```
Comment (nmathewson) [inline scripts/maint/pre-push.git-hook]: Should we be checking local_ref at all, or only remote_ref ?

Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: I suppose we should check both, to cover weird situations of local regular branch being pushed to remote upstream branch, and vice versa?

Comment (nmathewson) [reply scripts/maint/pre-push.git-hook]: In that case, I think we should explicitly check for match; pushing maint-0.3.5 to maint-0.3.4 is as bad as pushing bug12345 to maint-0.3.4.

Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: a94c3328829482e0da7a175e3dee089b9113d553 and 59c26154e80d589a069e54a9aa52eca2e14bf322 fixes that.

File: scripts/maint/pre-push.git-hook [hunk 4]
```diff
@@ -44,8 +44,9 @@ do
 			range="$remote_sha..$local_sha"
 		fi
 
-                if ref_is_upstream_branch "$local_ref" == 0 ||
-                        ref_is_upstream_branch "$remote_ref"  == 0
+                if (ref_is_upstream_branch "$local_ref" == 0 ||
+                        ref_is_upstream_branch "$remote_ref"  == 0) &&
+                        "$local_ref" ne "$remote_ref"
```
Comment (nmathewson) [inline scripts/maint/pre-push.git-hook]: Does this part need to be in [square brackets] ?

Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: Shellcheck does not complain about this, so I think not? Then again I'm really not sure.

Comment (rl1987) [reply scripts/maint/pre-push.git-hook]: On second thought, let's use `[...]` for consistency with previous code. See e28bbc8bd5847aba774b5ea38e7e98fbe5dd11fd.

**Pull request closed without merging**